qsc/grad_B_tensor.py

```diff
@@ -1177,3 +1190,233 @@ def calculate_grad_grad_B_tensor(self, two_ways=False):
+    every point along the axis (hence with nphi points) where x, y and z
+    are the standard cartesian coordinates for a given
+    near-axis radius r and a Boozer poloidal angle vartheta (not theta).
+
+    Args:
+      r: the near-axis radius
+      theta: the Boozer poloidal angle vartheta (= theta-N*phi)
+    '''
+    B_vector_cylindrical = self.Bfield_cylindrical(r,theta)
+    phi = self.phi
+
+    B_x = np.cos(phi) * B_vector_cylindrical[0] - np.sin(phi) * B_vector_cylindrical[1]
+    B_y = np.sin(phi) * B_vector_cylindrical[0] + np.cos(phi) * B_vector_cylindrical[1]
+    B_z = B_vector_cylindrical[2]
+
+    B_vector_cartesian = np.array([B_x, B_y, B_z])
+
+    return B_vector_cartesian
+
+def Bfield_gradient_cylindrical(self):
+    '''
+    Function to calculate the gradient of the magnetic field vector B=(B_R,B_phi,B_Z)
+    at every point along the axis (hence with nphi points) where R, phi and Z
+    are the standard cylindrical coordinates.
+    '''
+    
+    # Define auxiliary variables
+    t = self.tangent_cylindrical.transpose()
+    n = self.normal_cylindrical.transpose()
+    b = self.binormal_cylindrical.transpose()
+    Bbar = self.Bbar
+    B0 = self.B0
+    sG = self.sG
+    G0 = self.G0
+    X1c = self.X1c
+    X1s = self.X1s
+    Y1c = self.Y1c
+    Y1s = self.Y1s
+    d_l_d_varphi = self.d_l_d_varphi
+    curvature = self.curvature
+    torsion = self.torsion
+    iotaN = self.iotaN
+    d_X1c_d_varphi = self.d_X1c_d_varphi
+    d_X1s_d_varphi = self.d_X1s_d_varphi
+    d_Y1s_d_varphi = self.d_Y1s_d_varphi
+    d_Y1c_d_varphi = self.d_Y1c_d_varphi
+
+    factor = B0*B0/Bbar/d_l_d_varphi
+    grad_B_vector_nn = factor * (d_X1c_d_varphi * Y1s - d_X1s_d_varphi * Y1c + iotaN * (X1s * Y1s + X1c * Y1c))
+    grad_B_vector_bn = factor * (X1c * d_X1s_d_varphi - X1s * d_X1c_d_varphi - sG * Bbar * d_l_d_varphi * torsion / B0 - iotaN * (X1s * X1s + X1c * X1c))
+    grad_B_vector_nb = factor * (d_Y1c_d_varphi * Y1s - d_Y1s_d_varphi * Y1c + sG * Bbar * d_l_d_varphi * torsion / B0 + iotaN * (Y1s * Y1s + Y1c * Y1c))
+    grad_B_vector_bb = factor * (X1c * d_Y1s_d_varphi - X1s * d_Y1c_d_varphi - iotaN * (X1s * Y1s + X1c * Y1c))
+    grad_B_vector_tn = sG * B0 * curvature
+    grad_B_vector_nt = sG * B0 * curvature
+    grad_B_vector_tt = 0#sG * np.matmul(self.d_d_varphi,B0)/d_l_d_varphi
+
+    grad_B_vector_cylindrical = np.array([[
+                                  grad_B_vector_nn * n[i] * n[j] \
+                                + grad_B_vector_bn * b[i] * n[j] + grad_B_vector_nb * n[i] * b[j] \
+                                + grad_B_vector_bb * b[i] * b[j] \
+                                + grad_B_vector_tn * t[i] * n[j] + grad_B_vector_nt * n[i] * t[j] \
+                                + grad_B_vector_tt * t[i] * t[j]
+                                for i in range(3)] for j in range(3)])
+
+    return grad_B_vector_cylindrical
+
+def Bfield_gradient_cartesian(self):
```

Comment:
Consider renaming this function for consistency with the names `grad_B_tensor` and `grad_grad_B_tensor` used elsewhere in pyqsc.